README.md
![Maintenance](https://img.shields.io/badge/maintenance-experimental-blue.svg)
[![crates.io](https://img.shields.io/crates/v/ftdi-embedded-hal.svg)](https://crates.io/crates/ftdi-embedded-hal)
[![docs.rs](https://docs.rs/ftdi-embedded-hal/badge.svg)](https://docs.rs/ftdi-embedded-hal/)
[![Build Status](https://github.com/ftdi-rs/ftdi-embedded-hal/workflows/CI/badge.svg)](https://github.com/ftdi-rs/ftdi-embedded-hal/actions)

# ftdi-embedded-hal

This is an [embedded-hal] implementation for the FTDI chips
that can use various drivers including [libftd2xx] and [ftdi-rs].

This enables development of embedded device drivers without the use of
a microcontroller. The FTDI devices interface with PC via USB, and
provide a multi-protocol synchronous serial engine to interface
with most GPIO, SPI, I2C embedded devices.

**Note:**
This is strictly a development tool.
The crate contains runtime borrow checks and explicit panics to adapt the
FTDI device into the [embedded-hal] traits.

## Quickstart

* Enable the "libftd2xx-static" feature flag to use static linking with libftd2xx driver.
* Linux users only: Add [udev rules].

```toml
[dependencies.ftdi-embedded-hal]
version = "0.14.1"
features = ["libftd2xx", "libftd2xx-static"]
```

## Limitations

* Limited trait support: SPI, I2C, Delay, InputPin, and OutputPin traits are implemented.
* Limited device support: FT232H, FT2232H, FT4232H.
* Limited SPI modes support: MODE0, MODE2.

## Examples

### SPI

Communicate with SPI devices using [ftdi-rs] driver:
```rust
use ftdi_embedded_hal as hal;

let device = ftdi::find_by_vid_pid(0x0403, 0x6010)
    .interface(ftdi::Interface::A)
    .open()?;

let hal = hal::FtHal::init_freq(device, 3_000_000)?;
let spi = hal.spi()?;
```

Communicate with SPI devices using [libftd2xx] driver:
```rust
use ftdi_embedded_hal as hal;

let device = libftd2xx::Ft2232h::with_description("Dual RS232-HS A")?;

let hal = hal::FtHal::init_freq(device, 3_000_000)?;
let spi = hal.spi()?;
```

### I2C

Communicate with I2C devices using [ftdi-rs] driver:
```rust
use ftdi_embedded_hal as hal;

let device = ftdi::find_by_vid_pid(0x0403, 0x6010)
    .interface(ftdi::Interface::A)
    .open()?;

let hal = hal::FtHal::init_freq(device, 400_000)?;
let i2c = hal.i2c()?;
```

Communicate with I2C devices using [libftd2xx] driver:
```rust
use ftdi_embedded_hal as hal;

let device = libftd2xx::Ft232h::with_description("Single RS232-HS")?;

let hal = hal::FtHal::init_freq(device, 400_000)?;
let i2c = hal.i2c()?;
```

### GPIO

Control GPIO pins using [libftd2xx] driver:
```rust
use ftdi_embedded_hal as hal;

let device = libftd2xx::Ft232h::with_description("Single RS232-HS")?;

let hal = hal::FtHal::init_default(device)?;
let gpio = hal.ad6();
```

Control GPIO pins using [ftdi-rs] driver:
```rust
use ftdi_embedded_hal as hal;

let device = ftdi::find_by_vid_pid(0x0403, 0x6010)
    .interface(ftdi::Interface::A)
    .open()?;

let hal = hal::FtHal::init_default(device)?;
let gpio = hal.ad6();
```

### More examples

* [newAM/eeprom25aa02e48-rs]: read data from Microchip 25AA02E48 SPI EEPROM
* [newAM/bme280-rs]: read samples from Bosch BME280 sensor via I2C protocol

[embedded-hal]: https://github.com/rust-embedded/embedded-hal
[ftdi-rs]: https://github.com/tanriol/ftdi-rs
[libftd2xx crate]: https://github.com/ftdi-rs/libftd2xx-rs/
[libftd2xx]: https://github.com/ftdi-rs/libftd2xx-rs
[newAM/eeprom25aa02e48-rs]: https://github.com/newAM/eeprom25aa02e48-rs/blob/main/examples/ftdi.rs
[newAM/bme280-rs]: https://github.com/newAM/bme280-rs/blob/main/examples/ftdi-i2c.rs
[udev rules]: https://github.com/ftdi-rs/libftd2xx-rs/#udev-rules
[setup executable]: https://www.ftdichip.com/Drivers/CDM/CDM21228_Setup.zip


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.14.1] - 2023-05-10
### Fixed
- Changed the `Error` and `ErrorKind` types from private to public.

## [0.14.0] - 2022-11-27
### Changed
- Removed lifetimes on `OutputPin`, `InputPin`, `I2c`, `Spi`, and `SpiDevice` to improve ease-of-use.

## [0.13.0] - 2022-09-28
### Changed
- Updated the alpha release of `embedded-hal` from `1.0.0-alpha.8` to `1.0.0-alpha.9`.

## [0.12.0] - 2022-09-03
### Added
- Added re-exports for `libftd2xx` and `ftdi` when the respective feature is used.
- Added `embedded-hal` version `1.0.0-alpha.8` trait implementations for:
  - GPIOs
  - Delay
  - SPI

### Changed
- Changed the `embedded-hal` version `0.2` re-export name from `embedded-hal` to
  `eh0` to differentiate from `embedded-hal` version `1.0.0-alpha.8`.

## [0.11.0] - 2022-01-18
### Added
- Added support for input pins.

### Changed
- The `ad0` - `ad7` methods to get an `OutputPin` now return a `Result` to
  support input pins, previously these methods were infallible.

## [0.10.0] - 2021-11-08
### Added
- Added support for `libftdi1` as a backend.

### Changed
- Renamed to `ftdi-embedded-hal`.
- The `ftd2xx` backend is no longer enabled by default.
- Changed the error type to support multiple backends.
- Updated the edition from 2018 to 2021.

## [0.9.1] - 2021-08-10
### Fixed
- Call `close()` on Drop, allowing recovery from failures in `init()`.

## [0.9.0] - 2021-07-01
### Changed
- Updated the `libftd2xx` dependency from 0.29.0 to 0.31.0.

## [0.8.0] - 2021-05-29
### Changed
- Updated the `libftd2xx` dependency from 0.28.0 to 0.29.0.

## [0.7.0] - 2021-04-18
### Added
- Added checks for missing ACKs from the I2C slave.
  Missing ACKs will now return an `NakError` from the I2C traits.

### Changed
- Changed the default implementation of I2C traits to wait for a slave ACK
  before transmitting more bytes.  The previous behavior can be retained by
  calling `set_fast(true)`.

## [0.6.0] - 2021-04-10
### Added
- Added support for the FT4232H.
- Added support for the FT2232H.

### Changed
- Changed the default linking method on Linux to dynamic.
  Static linking can be enabled with the `static` feature flag.
- Changed the I2C pins to input (tri-state) when in idle mode.

### Fixed
- Fixed AD0 (SCL) pulling low when when I2C is first initialized.
- Fixed I2C AD0 & AD1 (SCL & SDA out) being pulled low when another OutputPin
  changed state.

## [0.5.1] - 2021-03-20
### Fixed
- Fixed the I2C `Read` trait not setting the read address bit.
- Fixed the I2C `Write` trait not driving SDA as an output when clocking data
  out.

## [0.5.0] - 2021-03-20
### Added
- Added checks for pin allocation, trying to take output pins 0-2 while using
  the SPI interface will now result in panic.
- Added I2C traits.
- Added `Debug` for interface structures.
- Added `with_serial_number` and `with_description` constructors.

### Changed
- Changed the FTDI MPSSE initialization to occur once globally for the device
  instead of inside the SPI device trait.
- Changed the `Delay` structure to contain dummy data for possible future use.
- Change the `Ft232hHal::with_ft` to `impl From<Ft232h> for Ft232hHal`.

### Removed
- Removed `Eq` and `PartialEq` traits on the `Delay` structure.

## [0.4.0] - 2021-03-05
### Added
- Added a `Delay` structure that implements the embedded-hal delay traits.

### Changed
- Updated `libftd2xx` dependency from 0.24.0 to 0.25.0.
  This updates the vendor library from 1.4.8 to 1.4.22 for Linux targets.
  This should fix timeout conditions that previously occurred when rapidly
  toggling GPIO pins.

## [0.3.0] - 2021-02-14
### Changed
- Improved latency for GPIOs

### Fixed
- Fixed the example code for `Ft232hHal::new`.
- Fixed pins 5, 6, 7 not being usable as outputs.

## [0.2.0] - 2020-09-13
### Added
- Added SPI non-blocking traits.

### Changed
- Updated to libftd2xx 0.17.0
- Updated to embedded-hal 0.2.4

## [0.1.0] - 2020-09-12
- Initial release

[Unreleased]: https://github.com/ftdi-rs/ftdi-embedded-hal/compare/v0.14.0...HEAD
[0.14.0]: https://github.com/ftdi-rs/ftdi-embedded-hal/compare/v0.13.0...v0.14.0

src/error.rs
use std::fmt;
use std::io;

/// Error type.
#[derive(Debug)]
pub enum Error<E: std::error::Error> {
    /// Errors internal to this crate.
    Hal(ErrorKind),
    /// IO error
    Io(io::Error),
    /// Backend-specific error
    Backend(E),
}
/// Internal HAL errors
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ErrorKind {
    InvalidParams,
    InvalidClock,
    BusBusy,
    I2cNoAck,
    GpioPinBusy,
    GpioInvalidPin,
    SpiModeNotSupported,
}

impl ErrorKind {
    fn as_str(&self) -> &str {
        match *self {
            ErrorKind::InvalidParams => "Invalid input params",
            ErrorKind::BusBusy => "Bus is busy",
            ErrorKind::InvalidClock => "Clock is not valid",
            ErrorKind::I2cNoAck => "No ACK from slave",
            ErrorKind::GpioPinBusy => "GPIO pin is already in use",
            ErrorKind::GpioInvalidPin => "No such GPIO pin",
            ErrorKind::SpiModeNotSupported => "Mode not supported",
        }
    }
}

impl<E: std::error::Error> fmt::Display for Error<E> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Error::Io(e) => e.fmt(f),
            Error::Backend(e) => fmt::Display::fmt(&e, f),
            Error::Hal(e) => write!(f, "A regular error occurred {:?}", e.as_str()),
        }
    }
}

impl<E: std::error::Error> std::error::Error for Error<E> {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            Error::Io(e) => e.source(),
            Error::Backend(e) => e.source(),
            Error::Hal(_) => None,
        }
    }
}

impl<E: std::error::Error> From<io::Error> for Error<E> {
    fn from(e: io::Error) -> Self {
        Error::Io(e)
    }
}

#[cfg(feature = "ftdi")]
impl From<ftdi::Error> for Error<ftdi::Error> {
    fn from(e: ftdi::Error) -> Self {
        Error::Backend(e)
    }
}

#[cfg(feature = "libftd2xx")]
impl From<libftd2xx::TimeoutError> for Error<libftd2xx::TimeoutError> {
    fn from(e: libftd2xx::TimeoutError) -> Self {
        Error::Backend(e)
    }
}


src/lib.rs
//! This is an [embedded-hal] implementation for the FTDI chips
//! that can use various drivers including [libftd2xx] and [ftdi-rs].
//!
//! This enables development of embedded device drivers without the use of
//! a microcontroller. The FTDI devices interface with PC via USB, and
//! provide a multi-protocol synchronous serial engine to interface
//! with most GPIO, SPI, I2C embedded devices.
//!
//! **Note:**
//! This is strictly a development tool.
//! The crate contains runtime borrow checks and explicit panics to adapt the
//! FTDI device into the [embedded-hal] traits.
//!
//! # Quickstart
//!
//! * Enable the "libftd2xx-static" feature flag to use static linking with libftd2xx driver.
//! * Linux users only: Add [udev rules].
//!
//! ```toml
//! [dependencies.ftdi-embedded-hal]
//! version = "0.14.1"
//! features = ["libftd2xx", "libftd2xx-static"]
//! ```
//!
//! # Limitations
//!
//! * Limited trait support: SPI, I2C, Delay, InputPin, and OutputPin traits are implemented.
//! * Limited device support: FT232H, FT2232H, FT4232H.
//! * Limited SPI modes support: MODE0, MODE2.
//!
//! # Examples
//!
//! ## SPI
//!
//! Communicate with SPI devices using [ftdi-rs] driver:
//! ```no_run
//! use ftdi_embedded_hal as hal;
//!
//! # #[cfg(feature = "ftdi")]
//! # {
//! let device = ftdi::find_by_vid_pid(0x0403, 0x6010)
//!     .interface(ftdi::Interface::A)
//!     .open()?;
//!
//! let hal = hal::FtHal::init_freq(device, 3_000_000)?;
//! let spi = hal.spi()?;
//! # }
//! # Ok::<(), std::boxed::Box<dyn std::error::Error>>(())
//! ```
//!
//! Communicate with SPI devices using [libftd2xx] driver:
//! ```no_run
//! use ftdi_embedded_hal as hal;
//!
//! # #[cfg(feature = "libftd2xx")]
//! # {
//! let device = libftd2xx::Ft2232h::with_description("Dual RS232-HS A")?;
//!
//! let hal = hal::FtHal::init_freq(device, 3_000_000)?;
//! let spi = hal.spi()?;
//! # }
//! # Ok::<(), std::boxed::Box<dyn std::error::Error>>(())
//! ```
//!
//! ## I2C
//!
//! Communicate with I2C devices using [ftdi-rs] driver:
//! ```no_run
//! use ftdi_embedded_hal as hal;
//!
//! # #[cfg(feature = "ftdi")]
//! # {
//! let device = ftdi::find_by_vid_pid(0x0403, 0x6010)
//!     .interface(ftdi::Interface::A)
//!     .open()?;
//!
//! let hal = hal::FtHal::init_freq(device, 400_000)?;
//! let i2c = hal.i2c()?;
//! # }
//! # Ok::<(), std::boxed::Box<dyn std::error::Error>>(())
//! ```
//!
//! Communicate with I2C devices using [libftd2xx] driver:
//! ```no_run
//! use ftdi_embedded_hal as hal;
//!
//! # #[cfg(feature = "libftd2xx")]
//! # {
//! let device = libftd2xx::Ft232h::with_description("Single RS232-HS")?;
//!
//! let hal = hal::FtHal::init_freq(device, 400_000)?;
//! let i2c = hal.i2c()?;
//! # }
//! # Ok::<(), std::boxed::Box<dyn std::error::Error>>(())
//! ```
//!
//! ## GPIO
//!
//! Control GPIO pins using [libftd2xx] driver:
//! ```no_run
//! use ftdi_embedded_hal as hal;
//!
//! # #[cfg(feature = "libftd2xx")]
//! # {
//! let device = libftd2xx::Ft232h::with_description("Single RS232-HS")?;
//!
//! let hal = hal::FtHal::init_default(device)?;
//! let gpio = hal.ad6();
//! # }
//! # Ok::<(), std::boxed::Box<dyn std::error::Error>>(())
//! ```
//!
//! Control GPIO pins using [ftdi-rs] driver:
//! ```no_run
//! use ftdi_embedded_hal as hal;
//!
//! # #[cfg(feature = "ftdi")]
//! # {
//! let device = ftdi::find_by_vid_pid(0x0403, 0x6010)
//!     .interface(ftdi::Interface::A)
//!     .open()?;
//!
//! let hal = hal::FtHal::init_default(device)?;
//! let gpio = hal.ad6();
//! # }
//! # Ok::<(), std::boxed::Box<dyn std::error::Error>>(())
//! ```
//!
//! ## More examples
//!
//! * [newAM/eeprom25aa02e48-rs]: read data from Microchip 25AA02E48 SPI EEPROM
//! * [newAM/bme280-rs]: read samples from Bosch BME280 sensor via I2C protocol
//!
//! [embedded-hal]: https://github.com/rust-embedded/embedded-hal
//! [ftdi-rs]: https://github.com/tanriol/ftdi-rs
//! [libftd2xx crate]: https://github.com/ftdi-rs/libftd2xx-rs/
//! [libftd2xx]: https://github.com/ftdi-rs/libftd2xx-rs
//! [newAM/eeprom25aa02e48-rs]: https://github.com/newAM/eeprom25aa02e48-rs/blob/main/examples/ftdi.rs
//! [newAM/bme280-rs]: https://github.com/newAM/bme280-rs/blob/main/examples/ftdi-i2c.rs
//! [udev rules]: https://github.com/ftdi-rs/libftd2xx-rs/#udev-rules
//! [setup executable]: https://www.ftdichip.com/Drivers/CDM/CDM21228_Setup.zip
#![forbid(missing_docs)]
#![forbid(unsafe_code)]

pub use eh0;
pub use eh1;
pub use ftdi_mpsse;

#[cfg(feature = "ftdi")]
pub use ftdi;

#[cfg(feature = "libftd2xx")]
pub use libftd2xx;

mod delay;
mod error;
mod gpio;
mod i2c;
mod spi;

pub use crate::error::{Error, ErrorKind};
pub use delay::Delay;
pub use gpio::{InputPin, OutputPin};
pub use i2c::I2c;
pub use spi::{Spi, SpiDevice};

use ftdi_mpsse::{MpsseCmdExecutor, MpsseSettings};
use std::sync::{Arc, Mutex};

/// State tracker for each pin on the FTDI chip.
#[derive(Debug, Clone, Copy)]
enum PinUse {
    I2c,
    Spi,
    Output,
    Input,
}

impl std::fmt::Display for PinUse {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PinUse::I2c => write!(f, "I2C"),
            PinUse::Spi => write!(f, "SPI"),
            PinUse::Output => write!(f, "OUTPUT"),
            PinUse::Input => write!(f, "INPUT"),
        }
    }
}

#[derive(Debug)]
struct FtInner<Device: MpsseCmdExecutor> {
    /// FTDI device.
    ft: Device,
    /// GPIO direction.
    direction: u8,
    /// GPIO value.
    value: u8,
    /// Pin allocation.
    pins: [Option<PinUse>; 8],
}

impl<Device: MpsseCmdExecutor> FtInner<Device> {
    /// Allocate a pin for a specific use.
    pub fn allocate_pin(&mut self, idx: u8, purpose: PinUse) {
        assert!(idx < 8, "Pin index {idx} is out of range 0 - 7");

        if let Some(current) = self.pins[usize::from(idx)] {
            panic!(
                "Unable to allocate pin {idx} for {purpose}, pin is already allocated for {current}"
            );
        } else {
            self.pins[usize::from(idx)] = Some(purpose)
        }
    }

    pub fn deallocate_pin(&mut self, idx: u8) {
        assert!(idx < 8, "Pin index {idx} is out of range 0 - 7");

        self.pins[usize::from(idx)].take();
    }
}

impl<Device: MpsseCmdExecutor> From<Device> for FtInner<Device> {
    fn from(ft: Device) -> Self {
        FtInner {
            ft,
            direction: 0x00,
            value: 0x00,
            pins: [None; 8],
        }
    }
}

/// FTxxx device.
#[derive(Debug)]
pub struct FtHal<Device: MpsseCmdExecutor> {
    mtx: Arc<Mutex<FtInner<Device>>>,
}

impl<Device, E> FtHal<Device>
where
    Device: MpsseCmdExecutor<Error = E>,
    E: std::error::Error,
    Error<E>: From<E>,
{
    /// Initialize the FTDI MPSSE with sane defaults.
    ///
    /// Default values:
    ///
    /// * Reset the FTDI device.
    /// * 4k USB transfer size.
    /// * 1s USB read timeout.
    /// * 1s USB write timeout.
    /// * 16ms latency timer.
    /// * 100kHz clock frequency.
    ///
    /// # Example
    ///
    /// ```no_run
    /// use ftdi_embedded_hal as hal;
    ///
    /// # #[cfg(feature = "libftd2xx")]
    /// # {
    /// let device = libftd2xx::Ft232h::with_description("Single RS232-HS")?;
    /// let hal = hal::FtHal::init_default(device)?;
    /// # }
    /// # Ok::<(), std::boxed::Box<dyn std::error::Error>>(())
    /// ```
    pub fn init_default(device: Device) -> Result<FtHal<Device>, Error<E>> {
        let settings: MpsseSettings = MpsseSettings {
            clock_frequency: Some(100_000),
            ..Default::default()
        };

        Ok(FtHal::init(device, &settings)?)
    }

    /// Initialize the FTDI MPSSE with sane defaults and custom frequency
    ///
    /// # Example
    ///
    /// ```no_run
    /// use ftdi_embedded_hal as hal;
    ///
    /// # #[cfg(feature = "libftd2xx")]
    /// # {
    /// let device = libftd2xx::Ft232h::with_description("Single RS232-HS")?;
    /// let hal = hal::FtHal::init_freq(device, 3_000_000)?;
    /// # }
    /// # Ok::<(), std::boxed::Box<dyn std::error::Error>>(())
    /// ```
    pub fn init_freq(device: Device, freq: u32) -> Result<FtHal<Device>, Error<E>> {
        let settings: MpsseSettings = MpsseSettings {
            clock_frequency: Some(freq),
            ..Default::default()
        };

        Ok(FtHal::init(device, &settings)?)
    }

    /// Initialize the FTDI MPSSE with custom values.
    ///
    /// **Note:** The `mask` field of [`MpsseSettings`] is ignored for this function.
    ///
    /// **Note:** The clock frequency will be 2/3 of the specified value when in
    /// I2C mode.
    ///
    /// # Panics
    ///
    /// Panics if the `clock_frequency` field of [`MpsseSettings`] is `None`.
    ///
    /// # Example
    ///
    /// ```no_run
    /// use ftdi_embedded_hal as hal;
    /// use ftdi_mpsse::MpsseSettings;
    /// use std::time::Duration;
    ///
    /// let mpsse = MpsseSettings {
    ///     reset: false,
    ///     in_transfer_size: 4096,
    ///     read_timeout: Duration::from_secs(5),
    ///     write_timeout: Duration::from_secs(5),
    ///     latency_timer: Duration::from_millis(32),
    ///     mask: 0x00,
    ///     clock_frequency: Some(400_000),
    /// };
    ///
    /// # #[cfg(feature = "libftd2xx")]
    /// # {
    /// let device = libftd2xx::Ft232h::with_description("Single RS232-HS")?;
    /// let hal = hal::FtHal::init(device, &mpsse)?;
    /// # }
    /// # Ok::<(), std::boxed::Box<dyn std::error::Error>>(())
    /// ```
    ///
    /// [`MpsseSettings`]: ftdi_mpsse::MpsseSettings
    pub fn init(mut device: Device, mpsse_settings: &MpsseSettings) -> Result<FtHal<Device>, E> {
        device.init(mpsse_settings)?;

        Ok(FtHal {
            mtx: Arc::new(Mutex::new(device.into())),
        })
    }
}

impl<Device, E> FtHal<Device>
where
    Device: MpsseCmdExecutor<Error = E>,
    E: std::error::Error,
    Error<E>: From<E>,
{
    /// Aquire the SPI peripheral for the FT232H.
    ///
    /// Pin assignments:
    /// * AD0 => SCK
    /// * AD1 => MOSI
    /// * AD2 => MISO
    ///
    /// # Panics
    ///
    /// Panics if pin 0, 1, or 2 are already in use.
    ///
    /// # Example
    ///
    /// ```no_run
    /// use ftdi_embedded_hal as hal;
    ///
    /// # #[cfg(feature = "libftd2xx")]
    /// # {
    /// let device = libftd2xx::Ft2232h::with_description("Dual RS232-HS A")?;
    /// let hal = hal::FtHal::init_freq(device, 3_000_000)?;
    /// let spi = hal.spi()?;
    /// # }
    /// # Ok::<(), std::boxed::Box<dyn std::error::Error>>(())
    /// ```
    pub fn spi(&self) -> Result<Spi<Device>, Error<E>> {
        Spi::new(self.mtx.clone())
    }

    /// Aquire the SPI peripheral with a chip select pin.
    ///
    /// This is specific to embedded-hal version 1.
    ///
    /// Pin assignments:
    /// * AD0 => SCK
    /// * AD1 => MOSI
    /// * AD2 => MISO
    ///
    /// # Panics
    ///
    /// Panics if pin 0, 1, 2 or the CS pin are already in use.
    ///
    /// # Example
    ///
    /// ```no_run
    /// use ftdi_embedded_hal as hal;
    ///
    /// # #[cfg(feature = "libftd2xx")]
    /// # {
    /// let device = libftd2xx::Ft2232h::with_description("Dual RS232-HS A")?;
    /// let hal = hal::FtHal::init_freq(device, 3_000_000)?;
    /// let spi = hal.spi_device(3)?;
    /// # }
    /// # Ok::<(), std::boxed::Box<dyn std::error::Error>>(())
    /// ```
    pub fn spi_device(&self, cs_idx: u8) -> Result<SpiDevice<Device>, Error<E>> {
        SpiDevice::new(self.mtx.clone(), cs_idx)
    }

    /// Aquire the I2C peripheral for the FT232H.
    ///
    /// Pin assignments:
    /// * AD0 => SCL
    /// * AD1 => SDA
    /// * AD2 => SDA
    ///
    /// Yes, AD1 and AD2 are both SDA.
    /// These pins must be shorted together for I2C operation.
    ///
    /// # Panics
    ///
    /// Panics if pin 0, 1, or 2 are already in use.
    ///
    /// # Example
    ///
    /// ```no_run
    /// use ftdi_embedded_hal as hal;
    ///
    /// # #[cfg(feature = "libftd2xx")]
    /// # {
    /// let device = libftd2xx::Ft2232h::with_description("Dual RS232-HS A")?;
    /// let hal = hal::FtHal::init_freq(device, 3_000_000)?;
    /// let i2c = hal.i2c()?;
    /// # }
    /// # Ok::<(), std::boxed::Box<dyn std::error::Error>>(())
    /// ```
    pub fn i2c(&self) -> Result<I2c<Device>, Error<E>> {
        I2c::new(self.mtx.clone())
    }

    /// Aquire the digital output pin 0 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn ad0(&self, initial: bool) -> Result<OutputPin<Device>, Error<E>> {
        OutputPin::new(self.mtx.clone(), 0, initial)
    }

    /// Aquire the digital input pin 0 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn adi0(&self) -> Result<InputPin<Device>, Error<E>> {
        InputPin::new(self.mtx.clone(), 0)
    }

    /// Aquire the digital output pin 1 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn ad1(&self, initial: bool) -> Result<OutputPin<Device>, Error<E>> {
        OutputPin::new(self.mtx.clone(), 1, initial)
    }

    /// Aquire the digital input pin 1 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn adi1(&self) -> Result<InputPin<Device>, Error<E>> {
        InputPin::new(self.mtx.clone(), 1)
    }

    /// Aquire the digital output pin 2 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn ad2(&self, initial: bool) -> Result<OutputPin<Device>, Error<E>> {
        OutputPin::new(self.mtx.clone(), 2, initial)
    }

    /// Aquire the digital input pin 2 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn adi2(&self) -> Result<InputPin<Device>, Error<E>> {
        InputPin::new(self.mtx.clone(), 2)
    }

    /// Aquire the digital output pin 3 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn ad3(&self, initial: bool) -> Result<OutputPin<Device>, Error<E>> {
        OutputPin::new(self.mtx.clone(), 3, initial)
    }

    /// Aquire the digital input pin 3 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn adi3(&self) -> Result<InputPin<Device>, Error<E>> {
        InputPin::new(self.mtx.clone(), 3)
    }

    /// Aquire the digital output pin 4 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn ad4(&self, initial: bool) -> Result<OutputPin<Device>, Error<E>> {
        OutputPin::new(self.mtx.clone(), 4, initial)
    }

    /// Aquire the digital input pin 4 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn adi4(&self) -> Result<InputPin<Device>, Error<E>> {
        InputPin::new(self.mtx.clone(), 4)
    }

    /// Aquire the digital output pin 5 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn ad5(&self, initial: bool) -> Result<OutputPin<Device>, Error<E>> {
        OutputPin::new(self.mtx.clone(), 5, initial)
    }

    /// Aquire the digital input pin 5 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn adi5(&self) -> Result<InputPin<Device>, Error<E>> {
        InputPin::new(self.mtx.clone(), 5)
    }

    /// Aquire the digital output pin 6 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn ad6(&self, initial: bool) -> Result<OutputPin<Device>, Error<E>> {
        OutputPin::new(self.mtx.clone(), 6, initial)
    }

    /// Aquire the digital input pin 6 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn adi6(&self) -> Result<InputPin<Device>, Error<E>> {
        InputPin::new(self.mtx.clone(), 6)
    }

    /// Aquire the digital output pin 7 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn ad7(&self, initial: bool) -> Result<OutputPin<Device>, Error<E>> {
        OutputPin::new(self.mtx.clone(), 7, initial)
    }

    /// Aquire the digital input pin 7 for the FT232H.
    ///
    /// # Panics
    ///
    /// Panics if the pin is already in-use.
    pub fn adi7(&self) -> Result<InputPin<Device>, Error<E>> {
        InputPin::new(self.mtx.clone(), 7)
    }
}
